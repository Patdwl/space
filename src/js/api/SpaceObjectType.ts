export enum SpaceObjectType {
  UNKNOWN = 0,
  PAYLOAD = 1,
  ROCKET_BODY = 2,
  DEBRIS = 3,
  SPECIAL = 4,
  RADAR_MEASUREMENT = 5,
  RADAR_TRACK = 6,
  RADAR_OBJECT = 7,
  BALLISTIC_MISSILE = 8,
  STAR = 9,
  INTERGOVERNMENTAL_ORGANIZATION = 10,
  SUBORBITAL_PAYLOAD_OPERATOR = 11,
  PAYLOAD_OWNER = 12,
  METEOROLOGICAL_ROCKET_LAUNCH_AGENCY_OR_MANUFACTURER = 13,
  PAYLOAD_MANUFACTURER = 14,
  LAUNCH_AGENCY = 15,
  LAUNCH_SITE = 16,
  LAUNCH_POSITION = 17,
  LAUNCH_FACILITY = 18,
  CONTROL_FACILITY = 19,
  GROUND_SENSOR_STATION = 20,
  OPTICAL = 21,
  MECHANICAL = 22,
  PHASED_ARRAY_RADAR = 23,
  OBSERVER = 24,
  BISTATIC_RADIO_TELESCOPE = 25,
  COUNTRY = 26,
  LAUNCH_VEHICLE_MANUFACTURER = 27,
  ENGINE_MANUFACTURER = 28,
}
