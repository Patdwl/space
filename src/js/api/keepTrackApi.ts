import { SatMath } from '../satMath/satMath';
import { KeepTrackApi, SatObject, SettingsManager } from './keepTrackTypes';
import { html } from './templateLiterals';

declare global {
  // eslint-disable-next-line no-unused-vars
  interface Window {
    settingsManager: SettingsManager;
    jQuery: unknown;
    $: unknown;
    gremlins: any;
    randomizer: any;
    keepTrackApi: KeepTrackApi;
    dataLayer: any; // For Google Tag Manager
    _numeric: any;
    satellite: SatMath;
    M: any;
  }
}

export const register = (params: { method: string; cbName: string; cb: any }) => {
  // If this is a valid callback
  if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
    // Add the callback
    keepTrackApi.callbacks[params.method].push({ name: params.cbName, cb: params.cb });
  } else {
    throw new Error(`Invalid callback "${params.method}"!`);
  }
};
export const unregister = (params: { method: string; cbName: string }) => {
  // If this is a valid callback
  if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
    for (let i = 0; i < keepTrackApi.callbacks[params.method].length; i++) {
      if (keepTrackApi.callbacks[params.method][i].name == params.cbName) {
        keepTrackApi.callbacks[params.method].splice(i, 1);
        return;
      }
    }
    // If we got this far, it means we couldn't find the callback
    throw new Error(`Callback "${params.cbName} not found"!`);
  } else {
    // Couldn't find the method
    throw new Error(`Invalid callback "${params.method}"!`);
  }
};
export const isThisJest = () => typeof process !== 'undefined';

export const keepTrackApi: KeepTrackApi = {
  html: html,
  register: register,
  unregister: unregister,
  callbacks: {
    selectSatData: [],
    updateSelectBox: [],
    onCruncherReady: [],
    onCruncherMessage: [],
    uiManagerInit: [],
    uiManagerOnReady: [],
    bottomMenuClick: [],
    hideSideMenus: [],
    nightToggle: [],
    orbitManagerInit: [],
    adviceReady: [],
    drawManagerLoadScene: [],
    drawOptionalScenery: [],
    updateLoop: [],
    rmbMenuActions: [],
    rightBtnMenuAdd: [],
    updateDateTime: [],
    uiManagerFinal: [],
    loadCatalog: [],
    resetSensor: [],
    setSensor: [],
    onDrawLoopComplete: [],
  },
  methods: {
    selectSatData: (sat: any, satId: number) => {
      keepTrackApi.programs.soundManager.play('whoosh');
      keepTrackApi.callbacks.selectSatData.forEach((cb: any) => cb.cb(sat, satId));
    },
    updateSelectBox: (sat: any) => {
      keepTrackApi.callbacks.updateSelectBox.forEach((cb: any) => cb.cb(sat));
    },
    onCruncherReady: () => {
      keepTrackApi.callbacks.onCruncherReady.forEach((cb: any) => cb.cb());
    },
    onCruncherMessage: () => {
      keepTrackApi.callbacks.onCruncherMessage.forEach((cb: any) => cb.cb());
    },
    uiManagerInit: () => {
      keepTrackApi.callbacks.uiManagerInit.forEach((cb: any) => cb.cb());
    },
    uiManagerOnReady: () => {
      keepTrackApi.callbacks.uiManagerOnReady.forEach((cb: any) => cb.cb());
    },
    bottomMenuClick: (iconName: string) => {
      keepTrackApi.programs.soundManager.play('genericBeep');
      keepTrackApi.callbacks.bottomMenuClick.forEach((cb: any) => cb.cb(iconName));
    },
    hideSideMenus: () => {
      keepTrackApi.callbacks.hideSideMenus.forEach((cb: any) => cb.cb());
    },
    nightToggle: (gl: any, nightTexture: any, texture: any) => {
      keepTrackApi.callbacks.nightToggle.forEach((cb: any) => cb.cb(gl, nightTexture, texture));
    },
    orbitManagerInit: () => {
      keepTrackApi.callbacks.orbitManagerInit.forEach((cb: any) => cb.cb());
    },
    adviceReady: () => {
      keepTrackApi.callbacks.adviceReady.forEach((cb: any) => cb.cb());
    },
    drawManagerLoadScene: () => {
      keepTrackApi.callbacks.drawManagerLoadScene.forEach((cb: any) => cb.cb());
    },
    drawOptionalScenery: () => {
      keepTrackApi.callbacks.drawOptionalScenery.forEach((cb: any) => cb.cb());
    },
    updateLoop: () => {
      keepTrackApi.callbacks.updateLoop.forEach((cb: any) => cb.cb());
    },
    rmbMenuActions: (menuName: string, satnum = -1) => {
      keepTrackApi.callbacks.rmbMenuActions.forEach((cb: any) => cb.cb(menuName, satnum));
    },
    rightBtnMenuAdd: () => {
      keepTrackApi.callbacks.rightBtnMenuAdd.forEach((cb: any) => cb.cb());
    },
    updateDateTime: (date: Date) => {
      keepTrackApi.callbacks.updateDateTime.forEach((cb: any) => cb.cb(date));
    },
    uiManagerFinal: () => {
      keepTrackApi.callbacks.uiManagerFinal.forEach((cb: any) => cb.cb());
    },
    loadCatalog: async (): Promise<SatObject[]> => keepTrackApi.callbacks.loadCatalog[0].cb(),
    resetSensor: () => {
      keepTrackApi.callbacks.resetSensor.forEach((cb: any) => cb.cb());
    },
    setSensor: (sensor: any, id: number) => {
      keepTrackApi.callbacks.setSensor.forEach((cb: any) => cb.cb(sensor, id));
    },
    onDrawLoopComplete: () => {
      keepTrackApi.callbacks.onDrawLoopComplete.forEach((cb: any) => cb.cb());
    },
  },
  programs: <any>{},
};

// First time we call this module we should make it available to the rest of the application
if (!window.keepTrackApi) window.keepTrackApi = keepTrackApi;
