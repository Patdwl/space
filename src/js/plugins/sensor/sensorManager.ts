/**
 * /////////////////////////////////////////////////////////////////////////////
 *
 * sensorManager.ts is the primary interface between the user interface and the
 * ground based sensors.
 *
 * http://keeptrack.space
 *
 * @Copyright (C) 2016-2021 Theodore Kruczek
 *
 * Original source code released by James Yoder at https://github.com/jeyoder/ThingsInSpace/
 * under the MIT License. Please reference http://keeptrack.space/license/thingsinspace.txt
 *
 * KeepTrack is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * KeepTrack is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with
 * KeepTrack. If not, see <http://www.gnu.org/licenses/>.
 *
 * /////////////////////////////////////////////////////////////////////////////
 */

import { spaceObjType2Str } from '@app/js/lib/spaceObjType2Str';
import $ from 'jquery';
import { keepTrackApi } from '../../api/keepTrackApi';
import { SensorManager, SensorObject } from '../../api/keepTrackTypes';
import { sensorList } from './sensorList';

// Add new callbacks to the list of callbacks in keepTrackApi
keepTrackApi.callbacks.setSensor = [];
keepTrackApi.methods.setSensor = (sensor: SensorObject, id: number) => {
  keepTrackApi.callbacks.setSensor.forEach((cb: any) => cb.cb(sensor, id));
};

keepTrackApi.callbacks.resetSensor = [];
keepTrackApi.methods.resetSensor = () => {
  keepTrackApi.callbacks.resetSensor.forEach((cb: any) => cb.cb());
};

const emptySensor: SensorObject = {
  observerGd: {
    lat: null,
    lon: 0,
    alt: 0,
  },
  alt: null,
  country: '',
  lat: null,
  lon: null,
  name: '',
  obsmaxaz: 0,
  obsmaxel: 0,
  obsmaxrange: 0,
  obsminaz: 0,
  obsminel: 0,
  obsminrange: 0,
  shortName: '',
  staticNum: 0,
  sun: '',
  volume: false,
  zoom: 'geo',
};

// NOTE: This doesn't account for sensorManager.selectedSensor
export const checkSensorSelected = () => sensorManager.currentSensor[0].lat != null;

export const setCurrentSensor = (sensor: SensorObject[] | null): void => {
  // TODO: This function is totally redundant to setSensor. There should be
  // ONE selectedSensor/currentSensor and it should be an array of selected sensors.
  if (sensor === null) {
    sensorManager.currentSensor[0] = emptySensor;
  } else if (sensor[0] != null) {
    sensorManager.currentSensor = sensor;
  } else if (sensor != null) {
    throw new Error('SensorManager.setCurrentSensor: sensor is not an array');
  }
};

export const sensorListLength = () => Object.values(sensorList).reduce((acc) => acc++, 0);

export const setSensor = (selectedSensor: SensorObject | string, staticNum: number) => { // NOSONAR
  try {
    localStorage.setItem('currentSensor', JSON.stringify([selectedSensor, staticNum]));
  } catch {
    console.warn('Sensor Manager: Unable to set current sensor - localStorage issue!');
  }

  // Run any callbacks
  const { satSet, satellite, objectManager } = keepTrackApi.programs;
  keepTrackApi.methods.setSensor(selectedSensor, staticNum);

  if (selectedSensor == null && staticNum == null) {
    sensorManager.sensorTitle = '';
    $('#sensor-selected').text(sensorManager.sensorTitle);
    return;
  }
  if (selectedSensor === 'SSN') {
    sensorManager.sensorTitle = 'All Space Surveillance Network Sensors';
    const filteredSensors = Object.values(sensorList).filter((sensor) => sensor.country === 'United States' || sensor.country === 'United Kingdom' || sensor.country === 'Norway');
    sendSensorToOtherPrograms(filteredSensors);
  } else if (selectedSensor === 'CapeCodMulti') {
    let multiSensor = [];
    multiSensor.push({
      lat: 41.754785,
      lon: -70.539151,
      alt: 0.060966,
      obsminaz: 347,
      obsmaxaz: 227,
      obsminel: 6,
      obsmaxel: 85,
      obsminrange: 150,
      obsmaxrange: 725,
      volume: false,
    });
    multiSensor.push({
      lat: 41.754785,
      lon: -70.539151,
      alt: 0.060966,
      obsminaz: 347,
      obsmaxaz: 227,
      obsminel: 3,
      obsmaxel: 85,
      obsminrange: 700,
      obsmaxrange: 2450,
      volume: false,
    });
    multiSensor.push({
      lat: 41.754785,
      lon: -70.539151,
      alt: 0.060966,
      obsminaz: 347,
      obsmaxaz: 227,
      obsminel: 3,
      obsmaxel: 85,
      obsminrange: 2200,
      obsmaxrange: 5556,
      volume: false,
    });
    satSet.satCruncher.postMessage({
      typ: 'sensor',
      setlatlong: true,
      sensor: multiSensor,
      multiSensor: true,
    });
    satellite.setobs([sensorManager.sensorList.COD]);
    objectManager.setSelectedSat(-1);
    satSet.setColorScheme(settingsManager.currentColorScheme, true);
    sensorManager.sensorTitle = 'Cape Cod Multi Fence Radar';
  } else if (selectedSensor === 'NATO-MW') {
    sensorManager.sensorTitle = 'All North American Aerospace Defense Command Sensors';
    const filteredSensors = Object.values(sensorList).filter(
      (sensor) =>
        // eslint-disable-next-line implicit-arrow-linebreak
        sensor == sensorManager.sensorList.BLE ||
        sensor == sensorManager.sensorList.CAV ||
        sensor == sensorManager.sensorList.CDN ||
        sensor == sensorManager.sensorList.COD ||
        sensor == sensorManager.sensorList.CLR ||
        sensor == sensorManager.sensorList.FYL ||
        sensor == sensorManager.sensorList.THL
    );
    sendSensorToOtherPrograms(filteredSensors);
  } else if (selectedSensor === 'RUS-ALL') {
    sensorManager.sensorTitle = 'All Russian Sensors';
    const filteredSensors = Object.values(sensorList).filter(
      (sensor) =>
        // eslint-disable-next-line implicit-arrow-linebreak
        sensor == sensorManager.sensorList.ARM ||
        sensor == sensorManager.sensorList.BAL ||
        sensor == sensorManager.sensorList.GAN ||
        sensor == sensorManager.sensorList.LEK ||
        sensor == sensorManager.sensorList.MIS ||
        sensor == sensorManager.sensorList.OLE ||
        sensor == sensorManager.sensorList.PEC ||
        sensor == sensorManager.sensorList.PIO
    );
    sendSensorToOtherPrograms(filteredSensors);
  } else if (selectedSensor === 'LEO-LABS') {
    sensorManager.sensorTitle = 'All LEO Labs Sensors';
    const filteredSensors = Object.values(sensorList).filter(
      (sensor) =>
        // eslint-disable-next-line implicit-arrow-linebreak
        sensor == sensorManager.sensorList.MSR || sensor == sensorManager.sensorList.PFISR || sensor == sensorManager.sensorList.KSR
    );
    sendSensorToOtherPrograms(filteredSensors);
  } else if (selectedSensor === 'MD-ALL') {
    sensorManager.sensorTitle = 'All Missile Defense Agency Sensors';
    const filteredSensors = Object.values(sensorList).filter(
      (sensor) =>
        // eslint-disable-next-line implicit-arrow-linebreak
        sensor == sensorManager.sensorList.COD ||
        sensor == sensorManager.sensorList.BLE ||
        sensor == sensorManager.sensorList.CLR ||
        sensor == sensorManager.sensorList.FYL ||
        sensor == sensorManager.sensorList.THL ||
        sensor == sensorManager.sensorList.HAR ||
        sensor == sensorManager.sensorList.QTR ||
        sensor == sensorManager.sensorList.KUR ||
        sensor == sensorManager.sensorList.SHA ||
        sensor == sensorManager.sensorList.KCS ||
        sensor == sensorManager.sensorList.SBX
    );
    sendSensorToOtherPrograms(filteredSensors);
  } else {
    for (const sensor in sensorList) {
      if (sensorList[sensor] == selectedSensor || (sensorList[sensor].staticNum === staticNum && typeof staticNum != 'undefined')) {
        // TODO: selectedSensor is redundant and should be merged with currentSensor (this will make bugs if you don't remove
        // references to selectedSensor)
        sensorManager.selectedSensor = sensorList[sensor];

        $('#sensor-info-title').html("<a class='iframe' href='" + sensorManager.selectedSensor.url + "'>" + sensorManager.selectedSensor.name + '</a>');
        new Promise(() => {
          $('a.iframe').colorbox({
            iframe: true,
            width: '80%',
            height: '80%',
            fastIframe: false,
            closeButton: false,
          });
        }).catch((err) => console.warn(err));

        $('#sensor-type').html(spaceObjType2Str(sensorManager.selectedSensor.type));
        $('#sensor-country').html(sensorManager.selectedSensor.country);
        sensorManager.sensorTitle = sensorManager.selectedSensor.name;

        // Send single sensor as an array to other programs
        sendSensorToOtherPrograms([sensorList[sensor]]);
      }
    }
  }
  $('#sensor-selected').text(sensorManager.sensorTitle);
};
export const drawFov = (sensor: SensorObject) => {
  switch (sensor.shortName) {
    case 'COD':
    case 'BLE':
    case 'CLR':
    case 'THL':
      keepTrackApi.programs.lineManager.create(
        'scan2',
        [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), sensor.obsminaz, sensor.obsminaz + 120, sensor.obsminel, sensor.obsmaxrange],
        'c'
      );
      keepTrackApi.programs.lineManager.create(
        'scan2',
        [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), sensor.obsminaz + 120, sensor.obsmaxaz, sensor.obsminel, sensor.obsmaxrange],
        'c'
      );
      break;
    case 'FYL':
      // TODO: Find actual face directions
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), 0, 120, sensor.obsminel, sensor.obsmaxrange], 'c');
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), 120, 240, sensor.obsminel, sensor.obsmaxrange], 'c');
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), 240, 0, sensor.obsminel, sensor.obsmaxrange], 'c');
      break;
    case 'CDN':
      // NOTE: This will be a bit more complicated later
      keepTrackApi.programs.lineManager.create(
        'scan2',
        [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), sensor.obsminaz, sensor.obsmaxaz, sensor.obsminel, sensor.obsmaxrange],
        'c'
      );
      break;
    default:
      console.debug('Sensor not found');
      break;
  }
};

export const sensorManager: SensorManager = {
  sensorList: sensorList,
  setSensor: setSensor,
  checkSensorSelected: checkSensorSelected,
  sensorListLength: sensorListLength,
  drawFov: drawFov,
  sensorTitle: '',
  setCurrentSensor: setCurrentSensor,
  curSensorPositon: [0, 0, 0],
  whichRadar: '',
  selectedSensor: <SensorObject>null,
  sensorListUS: [
    sensorList.COD,
    sensorList.BLE,
    sensorList.CAV,
    sensorList.CLR,
    sensorList.EGL,
    sensorList.FYL,
    sensorList.THL,
    sensorList.MIL,
    sensorList.ALT,
    sensorList.ASC,
    sensorList.CDN,
  ],
  currentSensor: [
    {
      observerGd: {
        lat: <number>null,
        lon: 0,
        alt: 0,
      },
      alt: 0,
      country: '',
      lat: 0,
      lon: 0,
      name: '',
      obsmaxaz: 0,
      obsmaxel: 0,
      obsmaxrange: 0,
      obsminaz: 0,
      obsminel: 0,
      obsminrange: 0,
      shortName: '',
      staticNum: 0,
      sun: '',
      volume: false,
      zoom: 'leo',
    },
  ],
  defaultSensor: <SensorObject[]>[
    {
      observerGd: {
        lat: null,
        lon: 0,
        alt: 0,
      },
    },
  ],
  currentSensorList: null,
  currentSensorMultiSensor: false,
  tempSensor: null,
  isLookanglesMenuOpen: false,
  isCustomSensorMenuOpen: false,
};

export const sendSensorToOtherPrograms = (filteredSensors: SensorObject[]) => {
  const { satSet, satellite, objectManager } = keepTrackApi.programs;
  satSet.satCruncher.postMessage({
    typ: 'sensor',
    setlatlong: true,
    sensor: filteredSensors,
    multiSensor: filteredSensors.length > 1,
  });
  setCurrentSensor(filteredSensors);
  satellite.setobs(filteredSensors);
  objectManager.setSelectedSat(-1);

  if (typeof settingsManager.currentColorScheme === 'undefined') throw new Error('settingsManager.currentColorScheme is undefined');
  satSet.setColorScheme(settingsManager.currentColorScheme, true);
};
