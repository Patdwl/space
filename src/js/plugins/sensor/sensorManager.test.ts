import { defaultSensor, keepTrackApiStubs } from '../../api/apiMocks';
import { keepTrackApi } from '../../api/externalApi';
import { SensorObject } from '../../api/keepTrack';
import * as sensorManager from './sensorManager';
keepTrackApi.programs = { ...keepTrackApi.programs, ...keepTrackApiStubs.programs };

declare const settingsManager;

describe('sensorManager.checkSensorSelected', () => {
  test('0', () => {
    sensorManager.setCurrentSensor(null);
    let result: any = sensorManager.checkSensorSelected();
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    sensorManager.setCurrentSensor(defaultSensor);
    let result: any = sensorManager.checkSensorSelected();
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('sensorManager.sensorListLength', () => {
  test('0', () => {
    let result: any = sensorManager.sensorListLength();
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('sensorManager.setCurrentSensor', () => {
  test('0', () => {
    let result: any = sensorManager.setCurrentSensor({
      alt: 50,
      beamwidth: 64,
      changeObjectInterval: 1,
      country: 'GB',
      lat: 1,
      linkAehf: false,
      linkWgs: false,
      lon: 4,
      name: 'Pierre Edouard',
      observerGd: { lat: 320, lon: 350, alt: 50 },
      obsmaxaz: 1,
      obsmaxel: -1.0,
      obsmaxrange: 520,
      obsminaz: -5.48,
      obsminel: -0.5,
      obsminrange: -5.48,
      shortName: 'test',
      staticNum: 1000,
      sun: 'v4.0.0-rc.4',
      type: 'array',
      url: 'https://twitter.com/path?abc',
      volume: true,
      zoom: 'leo',
    });
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let result: any = sensorManager.setCurrentSensor({
      alt: 1,
      beamwidth: 1000,
      changeObjectInterval: 100,
      country: 'FR',
      lat: 550,
      linkAehf: true,
      linkWgs: false,
      lon: 410,
      name: 'Michael',
      observerGd: { lat: 100, lon: 30, alt: 3 },
      obsmaxaz: -5.48,
      obsmaxel: 0.5,
      obsmaxrange: 100,
      obsminaz: -100,
      obsminel: 10.0,
      obsminrange: 0,
      shortName: 'test',
      staticNum: 1,
      sun: '1.0.0',
      type: 'array',
      url: 'http://www.croplands.org/account/confirm?t=',
      volume: true,
      zoom: 'geo',
    });
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let result: any = sensorManager.setCurrentSensor({
      alt: 3,
      beamwidth: 0.0,
      changeObjectInterval: -5.48,
      country: 'France',
      lat: 520,
      linkAehf: true,
      linkWgs: false,
      lon: 30,
      name: 'Anas',
      observerGd: { lat: 4, lon: 400, alt: 1 },
      obsmaxaz: -100,
      obsmaxel: -1.0,
      obsmaxrange: 90,
      obsminaz: -5.48,
      obsminel: 10.0,
      obsminrange: -100,
      shortName: 'test',
      staticNum: 1000,
      sun: '4.0.0-beta1\t',
      type: 'number',
      url: 'https://',
      volume: false,
      zoom: 'leo',
    });
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let result: any = sensorManager.setCurrentSensor({
      alt: 50,
      beamwidth: 0,
      changeObjectInterval: 100,
      country: 'US',
      lat: 520,
      linkAehf: false,
      linkWgs: true,
      lon: 550,
      name: 'Pierre Edouard',
      observerGd: { lat: 50, lon: 4, alt: 3 },
      obsmaxaz: 100,
      obsmaxel: 0.5,
      obsmaxrange: 320,
      obsminaz: 100,
      obsminel: -29.45,
      obsminrange: -100,
      shortName: 'fake',
      staticNum: 1000,
      sun: '4.0.0-beta1\t',
      type: 'number',
      url: 'https://twitter.com/path?abc',
      volume: true,
      zoom: 'geo',
    });
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    let result: any = sensorManager.setCurrentSensor({
      alt: 1,
      beamwidth: 40,
      changeObjectInterval: -100,
      country: 'France',
      lat: 70,
      linkAehf: false,
      linkWgs: false,
      lon: 380,
      name: 'Edmond',
      observerGd: { lat: 30, lon: 380, alt: 50 },
      obsmaxaz: 1,
      obsmaxel: 0.5,
      obsmaxrange: 520,
      obsminaz: 0,
      obsminel: 1.0,
      obsminrange: 100,
      shortName: '009008401',
      staticNum: 10,
      sun: 'v4.0.0-rc.4',
      type: 'object',
      url: 'https://accounts.google.com/o/oauth2/revoke?token=%s',
      volume: true,
      zoom: 'leo',
    });
    expect(result).toMatchSnapshot();
  });

  test('5', () => {
    let result: any = sensorManager.setCurrentSensor({
      alt: -Infinity,
      beamwidth: -Infinity,
      changeObjectInterval: -Infinity,
      country: '',
      lat: -Infinity,
      linkAehf: true,
      linkWgs: true,
      lon: -Infinity,
      name: '',
      observerGd: { lat: -Infinity, lon: -Infinity, alt: -Infinity },
      obsmaxaz: -Infinity,
      obsmaxel: -Infinity,
      obsmaxrange: -Infinity,
      obsminaz: -Infinity,
      obsminel: -Infinity,
      obsminrange: -Infinity,
      shortName: '',
      staticNum: -Infinity,
      sun: '',
      type: '',
      url: '',
      volume: false,
      zoom: '',
    });
    expect(result).toMatchSnapshot();
  });
});

describe('sensorManager.setSensor', () => {
  beforeAll(() => {
    settingsManager.currentColorScheme = {
      calculateColorBuffers: () => {},
    };
  });
  test('0', () => {
    let result: any = sensorManager.setSensor(
      {
        alt: 5,
        beamwidth: 48,
        changeObjectInterval: -100,
        country: 'United States',
        lat: 400,
        linkAehf: true,
        linkWgs: false,
        lon: 90,
        name: 'George',
        observerGd: { lat: 550, lon: 410, alt: 50 },
        obsmaxaz: -100,
        obsmaxel: -0.5,
        obsmaxrange: 1,
        obsminaz: 0,
        obsminel: 10.23,
        obsminrange: -5.48,
        shortName: 'test',
        staticNum: 1,
        sun: '1.0.0',
        type: 'object',
        url: 'https://accounts.google.com/o/oauth2/revoke?token=%s',
        volume: false,
        zoom: 'geo',
      },
      10
    );
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let result: any = sensorManager.setSensor('SSN', 1);
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let result: any = sensorManager.setSensor('CapeCodMulti', 1);
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let result: any = sensorManager.setSensor('NATO-MW', 1);
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    let result: any = sensorManager.setSensor('RUS-ALL', 1);
    expect(result).toMatchSnapshot();
  });

  test('5', () => {
    let result: any = sensorManager.setSensor('LEO-LABS', 1);
    expect(result).toMatchSnapshot();
  });

  test('6', () => {
    let result: any = sensorManager.setSensor('MD-ALL', 1);
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('sensorManager.drawFov', () => {
  test('0', () => {
    let result: any = sensorManager.drawFov(<SensorObject>{ shortName: 'COD' });
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let result: any = sensorManager.drawFov(<SensorObject>{ shortName: 'FYL' });
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let result: any = sensorManager.drawFov(<SensorObject>{ shortName: 'CDN' });
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let result: any = sensorManager.drawFov(<SensorObject>{ shortName: 'FAKE' });
    expect(result).toMatchSnapshot();
  });
});
