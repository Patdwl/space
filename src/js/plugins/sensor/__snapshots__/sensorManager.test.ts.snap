// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sensorManager.checkSensorSelected 0 1`] = `false`;

exports[`sensorManager.checkSensorSelected 1 1`] = `true`;

exports[`sensorManager.drawFov 0 1`] = `undefined`;

exports[`sensorManager.drawFov 1 1`] = `undefined`;

exports[`sensorManager.drawFov 2 1`] = `undefined`;

exports[`sensorManager.drawFov 3 1`] = `undefined`;

exports[`sensorManager.sensorListLength 0 1`] = `0`;

exports[`sensorManager.setCurrentSensor 0 1`] = `undefined`;

exports[`sensorManager.setCurrentSensor 1 1`] = `undefined`;

exports[`sensorManager.setCurrentSensor 2 1`] = `undefined`;

exports[`sensorManager.setCurrentSensor 3 1`] = `undefined`;

exports[`sensorManager.setCurrentSensor 4 1`] = `undefined`;

exports[`sensorManager.setCurrentSensor 5 1`] = `undefined`;

exports[`sensorManager.setSensor 0 1`] = `undefined`;

exports[`sensorManager.setSensor 1 1`] = `undefined`;

exports[`sensorManager.setSensor 2 1`] = `undefined`;

exports[`sensorManager.setSensor 3 1`] = `undefined`;

exports[`sensorManager.setSensor 4 1`] = `undefined`;

exports[`sensorManager.setSensor 5 1`] = `undefined`;

exports[`sensorManager.setSensor 6 1`] = `undefined`;
